src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -738,12 +895,164 @@ export class SavedObjectsRepository {
       id,
       type,
       updated_at: time,
-      version: encodeHitVersion(response),
+      version: encodeHitVersion(updateResponse),
+      ...(this._registry.isMultiNamespace(type) && {
+        namespaces: updateResponse.get._source.namespaces,
+      }),
       references,
       attributes,
     };
   }
 
+  /**
+   * Adds one or more namespaces to a given multi-namespace saved object. This method and
+   * [`deleteFromNamespaces`]{@link SavedObjectsRepository.deleteFromNamespaces} are the only ways to change which Spaces a multi-namespace
+   * saved object is shared to.
+   */
+  async addToNamespaces(
+    type: string,
+    id: string,
+    namespaces: string[],
+    options: SavedObjectsAddToNamespacesOptions = {}
+  ): Promise<{}> {
+    if (!this._allowedTypes.includes(type)) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    if (!this._registry.isMultiNamespace(type)) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(`${type} doesn't support namespaces`);
+    }
+
+    if (!Array.isArray(namespaces) || !namespaces.length) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(
+        'namespaces must be a non-empty array of strings'
+      );
+    }
+
+    const { version, namespace, refresh = DEFAULT_REFRESH_SETTING } = options;
+
+    const rawId = this._serializer.generateRawId(undefined, type, id);
+    const preflightResult = await this.preflightCheckIncludesNamespace(type, id, namespace);
+    const existingNamespaces = getSavedObjectNamespaces(undefined, preflightResult);
+    // there should never be a case where a multi-namespace object does not have any existing namespaces
+    // however, it is a possibility if someone manually modifies the document in Elasticsearch
+    const time = this._getCurrentTime();
+
+    const doc = {
+      updated_at: time,
+      namespaces: existingNamespaces ? unique(existingNamespaces.concat(namespaces)) : namespaces,
+    };
+
+    const updateResponse = await this._writeToCluster('update', {
+      id: rawId,
+      index: this.getIndexForType(type),
+      ...getExpectedVersionProperties(version, preflightResult),
+      refresh,
+      ignore: [404],
+      body: {
+        doc,
+      },
+    });
+
+    if (updateResponse.status === 404) {
+      // see "404s from missing index" above
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    return {};
+  }
+
+  /**
+   * Removes one or more namespaces from a given multi-namespace saved object. If no namespaces remain, the saved object is deleted
+   * entirely. This method and [`addToNamespaces`]{@link SavedObjectsRepository.addToNamespaces} are the only ways to change which Spaces a
+   * multi-namespace saved object is shared to.
+   */
+  async deleteFromNamespaces(
+    type: string,
```

Comment:
`If no namespaces remain, the saved object is deleted` -> Is that really a behavior we want? Shouldn't we be expecting to throw an error stating that it's not possible to remove every namespaces from a multi-NS object and indicate to use the `delete` method instead to perform deletion?

Reply:
No, since we don't want to change how consumers have to reason about saved objects, the `delete` operation simply removes a multi-namespace saved object from the current namespace. If a multi-namespace saved object exists in Space A and Space B, and you "delete" it from Space A, the repository actually updates the object and it will still exist in Space B.

However, if you remove all namespaces from a multi-namespace saved object -- if it winds up with no namespaces left for any reason, the object is deleted.

Reply:
Or, I see the reasoning. It makes sense, in a SO-api kind of way I guess 😄